registry.ts

```diff
@@ -42,53 +42,81 @@ export function defaultName(that: RegistryUrl): string {
   return n[1];
 }
 
-async function githubDownloadRelases(
-  owner: string,
-  repo: string,
-  lastVersion: string | undefined = undefined,
-): Promise<string[]> {
-  let url = `https://github.com/${owner}/${repo}/releases.atom`;
-  if (lastVersion) {
-    url += `?${lastVersion}`;
+// Link header format: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link
+function parseLinkHeader(res: Response): Record<string, string> {
+  const linkMap: Record<string, string> = {}
+  const headerValue = res.headers.get("link");
+  if (headerValue == null) {
+    return linkMap;
+  }
+
+  const links = headerValue.split(",");
+  for (const link of links) {
+    const match = link.match(/<(.+)>; rel="(.+)"/);
+    if (match) {
+      linkMap[match[2]] = match[1];
+    }
   }
-  // FIXME do we need to handle 404?
-
-  const page = await fetch(url);
-  const text = await page.text();
-  return [
-    ...text.matchAll(
-      /\<id\>tag\:github\.com\,2008\:Repository\/\d+\/(.*?)\<\/id\>/g,
-    ),
-  ].map((x) => x[1]);
+  return linkMap;
 }
 
 // export for testing purposes
 // FIXME this should really be lazy, we shouldn't always iterate everything...
-export const GR_CACHE: Map<string, string[]> = new Map<string, string[]>();
-async function githubReleases(
-  owner: string,
-  repo: string,
-  cache: Map<string, string[]> = GR_CACHE,
+export const CACHE: Map<string, string[]> = new Map<string, string[]>();
+async function fetchReleases(
+  url: string,
+  cacheKey: string,
+  cache: Map<string, string[]>
 ): Promise<string[]> {
-  const cacheKey = `${owner}/${repo}`;
   if (cache.has(cacheKey)) {
     return cache.get(cacheKey)!;
   }
-  const versions = await githubDownloadRelases(owner, repo);
-  if (versions.length === 10) {
-    let lastVersion: string | undefined = undefined;
-    // arbitrarily we're going to limit to 5 pages...?
-    let i: number = 0;
-    while (lastVersion !== versions[versions.length - 1] && i < 5) {
-      i++;
-      lastVersion = versions[versions.length];
-      versions.push(...await githubDownloadRelases(owner, repo, lastVersion));
-    }
+
+  // FIXME do we need to handle 404?
+  let res = await fetch(url);
+  let links = parseLinkHeader(res);
+  let tags: { name: string }[] = await res.json();
+  console.log(tags.length, "tags");
+  const versions = tags.map(tag => tag.name);
+
+  // arbitrarily fetch only 3 pages
+  let i = 0;
+  while (links.next && i < 3) {
+    res = await fetch(links.next);
+    tags = await res.json();
+    links = parseLinkHeader(res);
+    versions.push(...tags.map(tag => tag.name));
   }
+
   cache.set(cacheKey, versions);
   return versions;
 }
 
+function githubReleases(
+  owner: string,
+  repo: string,
+  cache: Map<string, string[]> = CACHE,
+): Promise<string[]> {
+  return fetchReleases(
+    `https://api.github.com/repos/${owner}/${repo}/tags`,
```

Comment:
The concern here is that this is rate limited (where the .atom / rss feed is not).